### ISSUE TITLE
[Proposal] Integrate Redis and MongoDB

### ISSUE THREAD

**Issue by microwavekonijn** [dignityofwar/digbot#47]
So this is kind of a large proposal as it requires some changes to make this happen. The start would be to include Redis and MongoDB and create a stack of these services and the Bot.

The reason is to include both is to persist data long term and make crashes less problematic. For instance when the bot crashes the list of people with a mute role gets wiped. MongoDB will be the store for server settings and things like user data, then we don't need the hardcoded settings for instance. Redis can be used as out jobs queue(like the play command), then when the bot crashes and restarts it can basically start where it left off(it can be possible that crashes can happen because of this data, which means when this data is used there should be a loose interpretation).

For MongoDB I know of a excellent ODM called Mongoose. The benefit choosing this over something like Mysql is that it will safe us some time short term and give us a more maintainable database.

Addition of these services would also require us to make some changes. Particularly to how we handle the configuration. For a start most of the configuration will just go from the files to the database and can be loaded in when the application starts. Other configurations like tokens, keys, connection details, etc will get default values which we will be able to overwrite using a special configuration file(for development) and environment variables. The environment variables will give us an easy way to link the different services to each other and have the possibility to overwrite them if necessary.

**Comment by JamesLongman** [comment]
I don't really have a lot of opinion over this as this isn't really my area of expertise. We definitely do need to have a group discussion about what we want to do database-wise when the time comes. Something like redis for persistent memory would be nice though.

We need to decide if we're going to continue work on the website, if so our database solution (previously SQL through API calls) will need to be integratable.

In any case this is pretty far into the future considering our turtle pace of development definitely postponed for v1+

**Comment by microwavekonijn** [comment]
Integration of Redis would be a great start and wouldn't be that difficult.

For the website, I am working on something that might be very interesting. However an api would be very advanced compared to just implementing a database. There will definitely be features that require this, but we will probably cross that bridge when we get there.

**Comment by JamesLongman** [comment]
Yeah I don't even know what's going on with the website repo. Unlike with a node.js project, I quite frankly do not have the ability to be taking any sort of leading role regarding that. To the best of my knowledge, Mael would like that repo moved as well but we will have to ask Llama what he wants to happen with that

**Comment by microwavekonijn** [comment]
So this is a proposal and I will try to explain as best as I can what I want to do.

### Service container
So this will be the heart of the bot. It will act as a register and based on that provide the right information and methods to whatever is requesting it. To give an example we can register a logger, when we need the logger we can ask the service container for that logger and then use it. The goal is to make something that is really flexible in use and allows for easy modification. The service container should in the end be able to resolve dependencies on it's own based on it registry.

### Service providers
The service providers are the ones filling the registry of the service container and then boot the appropriate services. This is basically splitting what we have now over multiple files creating a better overview of what is setup where and how.

### Bull & Redis
To immediately address of the issue of persistence I want to include Bull & Redis. This is basically to create a queues and we will most definitely benefit from this. Think of something like the play commands queue, or where it needs to process something that takes time(so the bot immediately response with "working on it" and then edits the message once it is done with the appropriate content, for instance the cat command when the api is down :P). In the service provider we can setup the queues for various jobs.
